// enkanetwork.js/dist/structs/Enka/EnkaProfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Profile = exports.EnkaProfile = void 0;
/**
 * A class that structures the profile data.
 */
class EnkaProfile {
    /**
     * The username of the profile.
     */
    username;
    /**
     * The profile info.
     */
    profile;
    /**
     * The ID of the profile.
     */
    id;
    /**
     * Creates a new `EnkaProfile` instance.
     * @param data - The data of the profile.
     */
    constructor(data) {
        this.username = data.username;
        this.profile = new Profile(data.profile);
        this.id = data.id;
    }
}
exports.EnkaProfile = EnkaProfile;
/**
 * A class to structure the profile info.
 */
class Profile {
    /**
     * The bio of the profile.
     */
    bio;
    /**
     * The level of the profile.
     */
    level;
    /**
     * The signup state of the profile.
     */
    signup_state;
    /**
     * The avatar of the profile.
     */
    avatar;
    /**
     * The url of the profile picture.
     */
    pfp_url;
    /**
     * Creates a new `Profile` instance.
     * @param data - The data of the profile's info.
     */
    constructor(data) {
        this.bio = data.bio;
        this.level = data.level;
        this.signup_state = data.signup_state;
        this.avatar = data.avatar || "";
        this.pfp_url = data.image_url || "";
    }
}
exports.Profile = Profile;
