// enkanetwork.js/dist/structs/Enka/HoyoBuilds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HoyoBuilds = void 0;
const client_1 = require("../../client");
const Genshin_1 = require("../Genshin");
const StarRail_1 = require("../StarRail");
/**
 * A class that structures the builds data of a hoyo.
 */
class HoyoBuilds {
    /**
     * The build's ID.
     */
    id;
    /**
     * The build's name.
     */
    name;
    /**
     * The build's character ID.
     */
    characterId;
    /**
     * The build's character name.
     */
    characterName;
    /**
     * The build's character info.
     */
    characterInfo;
    /**
     * The build's order.
     */
    order;
    /**
     * If the build is live.
     */
    live;
    /**
     * The build settings.
     */
    settings;
    /**
     * If the build is public.
     */
    isPublic;
    /**
     * The custom image of the build.
     */
    image;
    /**
     * The type of build (Genshin or StarRail).
     */
    type;
    /**
     * Creates a new `UserBuilds` instance.
     * @param data - The data of the user builds.
     * @param language - The language to get the names.
     */
    constructor(data, language) {
        const { genshin: genshinFinder, starrail: starrailFinder } = new client_1.AssetFinder();
        this.id = data.id;
        this.name = data.name;
        this.characterId = data.avatar_id;
        this.characterName =
            data.hoyo_type == 0
                ? genshinFinder.character(this.characterId).name
                : starrailFinder.character(this.characterId).name;
        this.characterInfo =
            data.hoyo_type == 0
                ? new Genshin_1.Characters(data.avatar_data, language)
                : new StarRail_1.SRCharacters(data.avatar_data, language);
        this.order = data.order;
        this.live = data.live;
        this.settings = data.settings;
        this.isPublic = data.public;
        this.image = data.image || "";
        this.type = data.hoyo_type;
    }
}
exports.HoyoBuilds = HoyoBuilds;
