// enkanetwork.js/dist/utils/Genshin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.theaterDifficulty = exports.substats = exports.weapons = exports.talents = exports.skills = exports.profilepictures = exports.namecards = exports.hashes = exports.costumes = exports.characters = void 0;
var characters_json_1 = require("./characters.json");
Object.defineProperty(exports, "characters", { enumerable: true, get: function () { return __importDefault(characters_json_1).default; } });
var costumes_json_1 = require("./costumes.json");
Object.defineProperty(exports, "costumes", { enumerable: true, get: function () { return __importDefault(costumes_json_1).default; } });
var hashes_json_1 = require("./hashes.json");
Object.defineProperty(exports, "hashes", { enumerable: true, get: function () { return __importDefault(hashes_json_1).default; } });
var namecards_json_1 = require("./namecards.json");
Object.defineProperty(exports, "namecards", { enumerable: true, get: function () { return __importDefault(namecards_json_1).default; } });
var profilepictures_json_1 = require("./profilepictures.json");
Object.defineProperty(exports, "profilepictures", { enumerable: true, get: function () { return __importDefault(profilepictures_json_1).default; } });
var skills_json_1 = require("./skills.json");
Object.defineProperty(exports, "skills", { enumerable: true, get: function () { return __importDefault(skills_json_1).default; } });
var talents_json_1 = require("./talents.json");
Object.defineProperty(exports, "talents", { enumerable: true, get: function () { return __importDefault(talents_json_1).default; } });
var weapons_json_1 = require("./weapons.json");
Object.defineProperty(exports, "weapons", { enumerable: true, get: function () { return __importDefault(weapons_json_1).default; } });
var substats_json_1 = require("./substats.json");
Object.defineProperty(exports, "substats", { enumerable: true, get: function () { return __importDefault(substats_json_1).default; } });
var theaterdifficulty_json_1 = require("./theaterdifficulty.json");
Object.defineProperty(exports, "theaterDifficulty", { enumerable: true, get: function () { return __importDefault(theaterdifficulty_json_1).default; } });
