// enkanetwork.js/dist/structs/AssetFinder/Genshin/Costume.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CostumeAssets = void 0;
const utils_1 = require("../../../utils");
const costumes = utils_1.costumes;
/**
 * A class that structures the costume assets.
 */
class CostumeAssets {
    /**
     * The assets of the costume.
     */
    assets;
    /**
     * Creates a new `Costume` instance.
     * @param costumeId - The ID of the costume.
     */
    constructor(costumeId) {
        const costume = costumes[costumeId];
        this.assets = costume
            ? new CostumeImages(costume)
            : {};
    }
}
exports.CostumeAssets = CostumeAssets;
/**
 * A class that structures the costume images.
 */
class CostumeImages {
    /**
     * The side icon of the costume.
     */
    sideIconName;
    /**
     * The icon of the costume.
     */
    icon;
    /**
     * The splash art of the costume.
     */
    art;
    /**
     * Creates a new `Costume` instance.
     * @param data - The data of the costume images.
     */
    constructor(data) {
        this.sideIconName = data ? data.sideIconName : "";
        this.icon = data ? data.icon : "";
        this.art = data ? data.art : "";
    }
}
