// enkanetwork.js/dist/structs/AssetFinder/Genshin/Namecard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamecardImages = exports.NamecardAssets = void 0;
const utils_1 = require("../../../utils");
const namecards = utils_1.namecards;
const hashes = utils_1.hashes;
/**
 * A class that structures the namecard assets and name.
 */
class NamecardAssets {
    /**
     * The name of the namecard.
     */
    name;
    /**
     * The assets of the namecard.
     */
    assets;
    /**
     * Creates a new `NamecardAssets` instance.
     * @param namecardId - The ID of the namecard.
     * @param language - The language to get the name.
     */
    constructor(namecardId, language) {
        const namecard = namecards[namecardId];
        this.name = hashes[language][namecard?.nameTextMapHash] || "";
        this.assets = namecard
            ? new NamecardImages(namecard)
            : {};
    }
}
exports.NamecardAssets = NamecardAssets;
/**
 * A class that structures the namecard images.
 */
class NamecardImages {
    /**
     * The namecard's icon.
     */
    icon;
    /**
     * The namecard's pic path.
     */
    picPath;
    /**
     * Creates a new `NamecardImages` instance.
     * @param data - The data of the namecard.
     */
    constructor(data) {
        this.icon = data ? data.icon : "";
        this.picPath = data ? data.picPath : [];
    }
}
exports.NamecardImages = NamecardImages;
