// enkanetwork.js/dist/structs/AssetFinder/Genshin/ProfilePicture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfilePictureImages = exports.ProfilePictureAssets = void 0;
const utils_1 = require("../../../utils");
const profilepictures = utils_1.profilepictures;
const hashes = utils_1.hashes;
/**
 * A class that structures the profile pictures assets and names
 */
class ProfilePictureAssets {
    /**
     * The name of the profile picture.
     */
    name;
    /**
     * The assets of the profile picture.
     */
    assets;
    /**
     * Creates a new `ProfilePictureAssets` instance.
     * @param profilePictureId - The ID of the profile picture.
     * @param language - The language to get the name.
     */
    constructor(profilePictureId, language) {
        const profilepicture = profilepictures[profilePictureId];
        this.name =
            hashes[language]?.[profilepicture?.nameTextMapHash] ||
                "";
        this.assets = profilepicture
            ? new ProfilePictureImages(profilepicture)
            : {};
    }
}
exports.ProfilePictureAssets = ProfilePictureAssets;
class ProfilePictureImages {
    /**
     * The icon of the profile picture.
     */
    icon;
    /**
     * The old format icon of the profile picture.
     */
    oldIcon;
    /**
     * Creates a new `ProfilePictureImages` instance.
     * @param data - The data of the profile picture.
     */
    constructor(data) {
        this.icon = data?.icon || "";
        this.oldIcon = data?.oldIcon || "";
    }
}
exports.ProfilePictureImages = ProfilePictureImages;
