// enkanetwork.js/dist/structs/AssetFinder/Genshin/Skill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkillImages = exports.SkillAssets = void 0;
const utils_1 = require("../../../utils");
const skills = utils_1.skills;
const hashes = utils_1.hashes;
/**
 * A class that structures the skill assets and names.
 */
class SkillAssets {
    /**
     * The skill name.
     */
    name;
    /**
     * The skill assets.
     */
    assets;
    /**
     * Creates a new `SkillAssets` instance.
     * @param skillId - The ID of the skill.
     * @param language - The language to get the name.
     */
    constructor(skillId, language) {
        this.name = hashes[language][skills[skillId]?.nameTextMapHash] || "";
        this.assets = skills[skillId]
            ? new SkillImages(skills[skillId])
            : {};
    }
}
exports.SkillAssets = SkillAssets;
/**
 * A class that structures the skill images.
 */
class SkillImages {
    /**
     * The skill's icon.
     */
    icon;
    /**
     * Creates a new `SkillImages` instance.
     * @param data - The data of the skill.
     */
    constructor(data) {
        this.icon = data ? data.icon : "";
    }
}
exports.SkillImages = SkillImages;
