// enkanetwork.js/dist/structs/AssetFinder/Genshin/Weapon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaponImages = exports.WeaponAssets = void 0;
const utils_1 = require("../../../utils");
const weapons = utils_1.weapons;
const hashes = utils_1.hashes;
/**
 * A class that structures the weapon assets and name.
 */
class WeaponAssets {
    /**
     * The weapon name.
     */
    name;
    /**
     * The weapon's rank stars.
     */
    stars;
    /**
     * The weapon assets.
     */
    assets;
    /**
     * Creates a new `WeaponAssets` instance.
     * @param weaponId - The ID of the weapon.
     * @param language - The language to get the name.
     */
    constructor(weaponId, language) {
        const weapon = weapons[weaponId];
        this.name = hashes[language][weapon?.nameTextMapHash] || "";
        this.stars = weapon?.rarity || 0;
        this.assets = weapon
            ? new WeaponImages(weapon)
            : {};
    }
}
exports.WeaponAssets = WeaponAssets;
/**
 * A class that structures the weapon images.
 */
class WeaponImages {
    /**
     * The weapon's icon.
     */
    icon;
    /**
     * The weapon's awaken icon
     */
    awakenIcon;
    /**
     * Creates a new `WeaponImages` instance.
     * @param data - The data of the weapon.
     */
    constructor(data) {
        this.icon = data ? data.icon : "";
        this.awakenIcon = data ? data.awakenIcon : "";
    }
}
exports.WeaponImages = WeaponImages;
