// enkanetwork.js/dist/structs/AssetFinder/StarRail/SRCharacter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SRCharacterSkills = exports.SRCharacterImages = exports.SRCharacterAssets = void 0;
const utils_1 = require("../../../utils");
const characters = utils_1.srcharacters;
const hashes = utils_1.srhashes;
const ranks = utils_1.ranks;
const skills = utils_1.srskills;
/**
 * A class that structures the character assets and name.
 */
class SRCharacterAssets {
    /**
     * The name of the character.
     */
    name;
    /**
     * The assets of the character.
     */
    assets;
    /**
     * Creates a new `SRCharacterAssets` instance.
     * @param characterId - The ID of the character.
     * @param language - The language to get the name.
     */
    constructor(characterId, language) {
        const character = characters[characterId];
        this.name = hashes[language]?.[character?.AvatarName] || "";
        this.assets = character
            ? new SRCharacterImages(character)
            : {};
    }
}
exports.SRCharacterAssets = SRCharacterAssets;
class SRCharacterImages {
    /**
     * The icon path of the character.
     */
    icon;
    /**
     * The gacha splash icon path of the character.
     */
    gachaIcon;
    /**
     * The paths of all the eidolon icons of the character.
     */
    eidolons;
    /**
     * The paths for the skill icons of the character.
     */
    skills;
    /**
     * Creates a new `SRCharacterImages` instance.
     * @param data - The data of the character assets.
     */
    constructor(data) {
        this.icon = data.AvatarSideIconPath;
        this.gachaIcon = data.AvatarCutinFrontImgPath;
        this.eidolons = data.RankIDList.map((id) => ranks[id].IconPath);
        this.skills = new SRCharacterSkills(data.SkillList);
    }
}
exports.SRCharacterImages = SRCharacterImages;
class SRCharacterSkills {
    /**
     * The path of the basic attack icon.
     */
    basicAttack;
    /**
     * The path of the skill icon.
     */
    skill;
    /**
     * The path of the ultimate icon.
     */
    ultimate;
    /**
     * The path of the talent icon.
     */
    talent;
    /**
     * The path of the technique icon.
     */
    technique;
    /**
     * Creates a new `SRCharacterSkills` instance.
     * @param data - The data of the character skills.
     */
    constructor(data) {
        this.basicAttack = skills[data[0]].SkillIcon;
        this.skill = skills[data[1]].SkillIcon;
        this.ultimate = skills[data[2]].SkillIcon;
        this.talent = skills[data[3]].SkillIcon;
        this.technique = skills[data[5]].SkillIcon;
    }
}
exports.SRCharacterSkills = SRCharacterSkills;
