// enkanetwork.js/dist/structs/AssetFinder/StarRail/SRRelic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SRRelic = void 0;
const utils_1 = require("../../../utils");
const relics = utils_1.relics;
const hashes = utils_1.srhashes;
class SRRelic {
    /**
     * The type of relic.
     */
    type;
    /**
     * The ID of the relic's set.
     */
    setId;
    /**
     * The path icon of the relic.
     */
    icon;
    /**
     * The name of the relic.
     */
    name;
    /**
     * The name of the relic's set.
     */
    setName;
    /**
     * Creates a new `SRRelic` instance.
     * @param relicId - The ID of the relic.
     * @param language - The language to get the names.
     */
    constructor(relicId, language) {
        this.type = relics[relicId]?.Type || "";
        this.setId = relics[relicId]?.SetID || 0;
        this.icon = relics[relicId]?.Icon || "";
        this.name = hashes[language]?.[relics[relicId]?.Name] || "";
        this.setName = hashes[language]?.[relics[relicId]?.SetName] || "";
    }
}
exports.SRRelic = SRRelic;
