// enkanetwork.js/dist/structs/StarRail/SRPlayerData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SRPlayerData = void 0;
const Enka_1 = require("../Enka");
const SRCharacters_1 = require("./SRCharacters");
const SRPlayer_1 = require("./SRPlayer");
/**
 * A class that structures the overall player's data.
 */
class SRPlayerData {
    /**
     * The information about the player.
     */
    player;
    /**
     * The information about the characters of the player.
     */
    characters;
    /**
     * The time in seconds until the profile data is refreshed.
     */
    ttl;
    /**
     * The Enka profile data that belongs the UID.
     */
    owner;
    /**
     * The UID of the player.
     */
    uid;
    /**
     * Creates a new `SRPlayerData` instance.
     * @param data - The data of the overall player data.
     * @param language - The language to get the names.
     */
    constructor(data, language) {
        this.player = data.detailInfo
            ? new SRPlayer_1.SRPlayer(data.detailInfo)
            : {};
        this.characters =
            data.detailInfo.avatarDetailList &&
                data.detailInfo.avatarDetailList.length > 0
                ? data.detailInfo.avatarDetailList.map((data) => new SRCharacters_1.SRCharacters(data, language))
                : [];
        this.ttl = data.ttl;
        this.owner = data.owner ? new Enka_1.Owner(data.owner) : {};
        this.uid = data.uid;
    }
}
exports.SRPlayerData = SRPlayerData;
