// enkanetwork.js/dist/structs/Genshin/Artifact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Artifact = void 0;
const client_1 = require("../../client");
const utils_1 = require("../../utils");
const substats = utils_1.substats;
/**
 * A class that structures the artifact's data.
 */
class Artifact {
    /**
     * The artifact's ID.
     */
    artifactId;
    /**
     * The artifact's level.
     */
    level;
    /**
     * The artifact's name text map hash.
     */
    nameTextMapHash;
    /**
     * The artifact's set name text map hash.
     */
    setNameTextMapHash;
    /**
     * The artifact's rank stars.
     */
    stars;
    /**
     * The artifact's main stat data.
     */
    mainstat;
    /**
     * The artifact's sub stats data.
     */
    substats;
    /**
     * The artifact's rolled sub stats IDs.
     */
    rolledSubstatsIds;
    /**
     * The roll quality of the substats.
     */
    substatsRollQuality;
    /**
     * The item type.
     */
    itemType;
    /**
     * The artifact's icon.
     */
    icon;
    /**
     * The artifact's equip type.
     */
    equipType;
    /**
     * The artifact's name.
     */
    name;
    /**
     * The artifact's set name.
     */
    setName;
    /**
     * Creates a new `Artifact` instance.
     * @param data - The data of the artifact.
     * @param language - The language to get the name.
     */
    constructor(data, language) {
        const { genshin: genshinFinder } = new client_1.AssetFinder({ language });
        this.artifactId = data.itemId;
        this.level = data.reliquary.level;
        this.nameTextMapHash = data.flat.nameTextMapHash;
        this.setNameTextMapHash = data.flat.setNameTextMapHash;
        this.stars = data.flat.rankLevel;
        this.mainstat = new ArtifactMainStat(data);
        this.substats = data.flat.reliquarySubstats
            ? data.flat.reliquarySubstats.map((data) => new ArtifactSubstats(data))
            : [];
        this.rolledSubstatsIds = data.reliquary.appendPropIdList;
        this.substatsRollQuality = data.reliquary.appendPropIdList
            ? data.reliquary.appendPropIdList.map((data) => new ArtifactRollQuality(data))
            : [];
        this.itemType = data.flat.itemType;
        this.icon = data.flat.icon;
        this.equipType = data.flat.equipType;
        this.name = genshinFinder.hash(this.nameTextMapHash).value;
        this.setName = genshinFinder.hash(this.setNameTextMapHash).value;
    }
    sumRollQuality() {
        const rollsQuality = [];
        if (this.substatsRollQuality.length > 0) {
            this.substatsRollQuality.forEach((substat) => {
                if (!rollsQuality.find((data) => data.type == substat.type)) {
                    rollsQuality.push({
                        type: substat.type,
                        rollQuality: this.substatsRollQuality
                            .filter((x) => x.type == substat.type)
                            .map((data) => data.rollQuality)
                    });
                }
            });
        }
        return rollsQuality;
    }
}
exports.Artifact = Artifact;
/**
 * A class that structures the artifact's main stat data.
 */
class ArtifactMainStat {
    /**
     * The artifact's main stat.
     */
    stat;
    /**
     * The artifact's main stat value.
     */
    statValue;
    /**
     * The artifact's main stat id.
     */
    id;
    /**
     * Creates a new `ArtifactMainStat` instance.
     * @param data - The data of the artifact.
     */
    constructor(data) {
        this.stat = data.flat.reliquaryMainstat.mainPropId;
        this.statValue = data.flat.reliquaryMainstat.statValue;
        this.id = data.reliquary.mainPropId;
    }
}
/**
 * A class that structures the artifact's sub stats data.
 */
class ArtifactSubstats {
    /**
     * The artifact's sub stat.
     */
    stat;
    /**
     * The artifact's sub stat value.
     */
    statValue;
    /**
     * Creates a new `ArtifactSubstats` instance.
     * @param data - The data of the artifact's substats.
     */
    constructor(data) {
        this.stat = data.appendPropId;
        this.statValue = data.statValue;
    }
}
class ArtifactRollQuality {
    /**
     * The ID of the substat roll.
     */
    id;
    /**
     * The type of the substat.
     */
    type;
    /**
     * The quality of the roll.
     * - 4 - Max roll | 3 - Almost max roll | 2 - Medium roll | 1 - Low roll
     */
    rollQuality;
    /**
     * Creates a new `ArtifactRollQuality` instance.
     * @param data - The rolled substat IDs.
     */
    constructor(substatId) {
        this.id = substatId;
        this.type = substats[substatId].propType;
        this.rollQuality = substatId % 10;
    }
}
