// enkanetwork.js/dist/structs/Genshin/Properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Properties = void 0;
/**
 * A class that structures the properties of the character.
 */
class Properties {
    /**
     * The xp of the character.
     */
    xp;
    /**
     * The ascension of the character.
     */
    ascension;
    /**
     * The level of the character.
     */
    level;
    /**
     * The stamina of the character.
     */
    stamina;
    /**
     * Creates a new `Properties` instance.
     * @param data - The data of the character's properties.
     */
    constructor(data) {
        this.xp = new PropertyContent(data[1001]);
        this.ascension = new PropertyContent(data[1002]);
        this.level = new PropertyContent(data[4001]);
        this.stamina = new PropertyContent(data[10010]);
    }
}
exports.Properties = Properties;
/**
 * A class that structures the content of the character's properties.
 */
class PropertyContent {
    /**
     * The type of propertie.
     */
    type;
    /**
     * The value of the propertie.
     */
    val;
    /**
     * Creates a new `PropertyContent` instance.
     * @param data - The data of the propertie content.
     */
    constructor(data) {
        this.type = data.type;
        this.val = data.val || "";
    }
}
