// enkanetwork.js/dist/structs/Genshin/Skills.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skills = void 0;
const client_1 = require("../../client");
const utils_1 = require("../../utils");
const characters = utils_1.characters;
/**
 * A class that structures the skill types data.
 */
class Skills {
    /**
     * The normal attacks data.
     */
    normalAttacks;
    /**
     * The elemental skill data.
     */
    elementalSkill;
    /**
     * The elemental burst data.
     */
    elementalBurst;
    /**
     * Creates a new `Skills` instance.
     * @param data - The data of the skill.
     * @param characterId - The ID of the character.
     * @param language - The language to get the names.
     */
    constructor(data, characterId, skillDepotId, language) {
        const idArr = Object.keys(data);
        if (["10000007", "10000005"].includes(characterId.toString()))
            characterId = `${characterId}-${skillDepotId}`;
        const character = characters[characterId];
        this.normalAttacks = new Skill(data[idArr[idArr.indexOf(character.skillOrder[0])]], +idArr[idArr.indexOf(character.skillOrder[0])], language);
        this.elementalSkill = idArr[1]
            ? new Skill(data[idArr[idArr.indexOf(character.skillOrder[1])]], +idArr[idArr.indexOf(character.skillOrder[1])], language)
            : {};
        this.elementalBurst = idArr[2]
            ? new Skill(data[idArr[idArr.indexOf(character.skillOrder[2])]], +idArr[idArr.indexOf(character.skillOrder[2])], language)
            : {};
    }
}
exports.Skills = Skills;
/**
 * A class that structures the skills data.
 */
class Skill {
    /**
     * The skill's level.
     */
    level;
    /**
     * The skill's ID.
     */
    id;
    /**
     * The skill's assets.
     */
    assets;
    /**
     * The skill's name.
     */
    name;
    /**
     * Creates a new `Skill` instance.
     * @param level - The level of the skill.
     * @param id - The ID of the skill.
     * @param language - The language to get the name.
     */
    constructor(level, id, language) {
        const { genshin: genshinFinder } = new client_1.AssetFinder({ language });
        const skill = genshinFinder.skill(id);
        this.level = level;
        this.id = id;
        this.assets = skill.assets;
        this.name = skill.name;
    }
}
