// enkanetwork.js/dist/structs/Genshin/Substats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Substats = void 0;
const utils_1 = require("../../utils");
const utils_2 = require("../../utils");
const substats = utils_1.substats;
const hashes = utils_2.hashes;
/**
 * A class that structures the parsed substats
 */
class Substats {
    /**
     * The type of the substat.
     */
    propType;
    /**
     * The value of the substat.
     */
    propValue;
    /**
     * The parsed name of the substat.
     */
    parsedPropType;
    /**
     * The parsed value of the substat.
     */
    parsedPropValue;
    /**
     * Indicates if the substat is percent or an integer.
     */
    isPercent;
    /**
     * Creates a new `Substats` instance.
     * @param substatId - The ID of the substat.
     */
    constructor(substatId, language) {
        const substat = substats[substatId];
        const split = substat?.propType.split("_") || [];
        const isPercent = ["HURT", "CRITICAL", "EFFICIENCY", "PERCENT", "ADD"].includes(split[split.length - 1] || "") || false;
        this.propType = substat?.propType || "";
        this.propValue = substat?.propValue || 0;
        this.parsedPropType =
            hashes[language]?.[substat?.propType] || "";
        this.parsedPropValue = substat?.propValue
            ? isPercent
                ? Math.round((substat.propValue * 100 + 0.0001) * 10) / 10
                : Math.round(substat.propValue)
            : 0;
        this.isPercent = isPercent;
    }
}
exports.Substats = Substats;
