// enkanetwork.js/dist/structs/Genshin/Weapon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Weapon = void 0;
const client_1 = require("../../client");
/**
 * A class that structures the weapon's data.
 */
class Weapon {
    /**
     * The weapon's ID.
     */
    weaponId;
    /**
     * The weapon's level.
     */
    level;
    /**
     * The weapon's ascension level.
     */
    ascensionLevel;
    /**
     * The weapon's refinement level.
     */
    refinement;
    /**
     * The weapon's name text map hash.
     */
    nameTextMapHash;
    /**
     * The weapon's rank stars.
     */
    stars;
    /**
     * the weapon's stats.
     */
    weaponStats;
    /**
     * The item type.
     */
    itemType;
    /**
     * The weapon's assets.
     */
    assets;
    /**
     * The weapon's name.
     */
    name;
    /**
     * Creates a new `Weapon` instance.
     * @param data - The data of the weapon.
     * @param language - The language to get the name.
     */
    constructor(data, language) {
        const { genshin: genshinFinder } = new client_1.AssetFinder({ language });
        this.weaponId = data.itemId;
        this.level = data.weapon.level;
        this.ascensionLevel = data.weapon.promoteLevel || "";
        this.refinement = new Refinement(data.weapon.affixMap);
        this.nameTextMapHash = data.flat.nameTextMapHash;
        this.stars = data.flat.rankLevel;
        this.weaponStats = data.flat.weaponStats.map((data) => new WeaponStats(data));
        this.itemType = data.flat.itemType;
        this.assets = genshinFinder.weapon(this.weaponId).assets;
        this.name = genshinFinder.weapon(this.weaponId).name;
    }
}
exports.Weapon = Weapon;
/**
 * A class that structures the weapon stats.
 */
class WeaponStats {
    /**
     * The weapon stat.
     */
    stat;
    /**
     * The weapon stat value.
     */
    statValue;
    /**
     * Creates a new `WeaponStats` instance.
     * @param data - The data of the weapon stats.
     */
    constructor(data) {
        this.stat = data.appendPropId;
        this.statValue = data.statValue;
    }
}
/**
 * A class that structures the weapon's refinement data.
 */
class Refinement {
    /**
     * The id of the refinement.
     */
    id;
    /**
     * The weapon's refinement level.
     */
    level;
    constructor(data) {
        this.id = data ? +Object.keys(data)[0] : "";
        this.level = data ? data[Object.keys(data)[0]] : "";
    }
}
