// enkanetwork.js/dist/utils/StarRail/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = exports.substattypes = exports.skilltree = exports.srskills = exports.relics = exports.ranks = exports.meta = exports.lightcones = exports.icons = exports.srhashes = exports.srcharacters = void 0;
var characters_json_1 = require("./characters.json");
Object.defineProperty(exports, "srcharacters", { enumerable: true, get: function () { return __importDefault(characters_json_1).default; } });
var hashes_json_1 = require("./hashes.json");
Object.defineProperty(exports, "srhashes", { enumerable: true, get: function () { return __importDefault(hashes_json_1).default; } });
var icons_json_1 = require("./icons.json");
Object.defineProperty(exports, "icons", { enumerable: true, get: function () { return __importDefault(icons_json_1).default; } });
var lightcones_json_1 = require("./lightcones.json");
Object.defineProperty(exports, "lightcones", { enumerable: true, get: function () { return __importDefault(lightcones_json_1).default; } });
var meta_json_1 = require("./meta.json");
Object.defineProperty(exports, "meta", { enumerable: true, get: function () { return __importDefault(meta_json_1).default; } });
var ranks_json_1 = require("./ranks.json");
Object.defineProperty(exports, "ranks", { enumerable: true, get: function () { return __importDefault(ranks_json_1).default; } });
var relics_json_1 = require("./relics.json");
Object.defineProperty(exports, "relics", { enumerable: true, get: function () { return __importDefault(relics_json_1).default; } });
var skills_json_1 = require("./skills.json");
Object.defineProperty(exports, "srskills", { enumerable: true, get: function () { return __importDefault(skills_json_1).default; } });
var skilltree_json_1 = require("./skilltree.json");
Object.defineProperty(exports, "skilltree", { enumerable: true, get: function () { return __importDefault(skilltree_json_1).default; } });
var substattypes_json_1 = require("./substattypes.json");
Object.defineProperty(exports, "substattypes", { enumerable: true, get: function () { return __importDefault(substattypes_json_1).default; } });
var types_json_1 = require("./types.json");
Object.defineProperty(exports, "types", { enumerable: true, get: function () { return __importDefault(types_json_1).default; } });
